api/routes_test.go

```diff
@@ -0,0 +1,128 @@
+package api
+
+import (
+	"encoding/base64"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/SkynetLabs/skynet-accounts/jwt"
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/fastrand"
+)
+
+// TestAPIKeyFromRequest ensures that apiKeyFromRequest works as expected.
+func TestAPIKeyFromRequest(t *testing.T) {
+	req := &http.Request{
+		Form:   make(map[string][]string),
+		Header: make(map[string][]string),
+	}
+
+	// API key from request with no API key.
+	_, err := apiKeyFromRequest(req)
+	if err != ErrNoAPIKey {
+		t.Fatalf("Expected '%s', got '%s'.", ErrNoAPIKey, err)
+	}
+
+	// API key from request form.
+	token := t.Name()
+	req.Form.Add("api_key", token)
+	tk, err := apiKeyFromRequest(req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk != token {
+		t.Fatalf("Expected '%s', got '%s'.", token, tk)
+	}
+
+	// API key from headers. Expect this to take precedence over request form.
+	token2 := t.Name() + "2"
+	req.Header.Set(APIKeyHeader, token2)
+	tk, err = apiKeyFromRequest(req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk == token {
+		t.Fatal("Form token took precedence over headers token.")
+	}
+	if tk != token2 {
+		t.Fatalf("Expected '%s', got '%s'.", token2, tk)
+	}
+}
+
+// TestTokenFromRequest ensures that tokenFromRequest works as expected.
+func TestTokenFromRequest(t *testing.T) {
+	// FIXME This is a weird and ugly workaround for the fact that I don't know
+	// 	how to refer to a file relative to the project's root dir.
+	jwt.AccountsJWKSFile = "../jwt/fixtures/jwks.json"
```

Comment:
this is how we do it in skyd fwiw. we have a whole `compatibility` directory of files for tests and the tests just reference them with relative paths.
you could make a const for the `fixtures` directory path if it makes you feel better.

Reply:
Yep this is fine. If you want to be safe that it works on all OSs you could use `filepath.Join` although we don't do that everywhere in Sia either. Go might just magically make it work on Windows or maybe Windows itself does even with the `/`.

Reply:
I'm not sure I'm 100% happy with this staying the final solution but I *am* happy that we're thinking alike.